std/typecons.d

```diff
@@ -7093,3 +7094,535 @@ unittest
         ubyte, ubyte, T3, T3,
     );
 }
+
+private enum isNullable(T) = is(typeof(T.init is null));
+private enum isCopyable(T) = is(typeof({ T t = T.init; auto copy = t; }));
+
+/**
+ * Range type to eliminate null dereferences.
+ *
+ * This type has two states; an $(I empty) state and a $(I non-empty) state.
+ * It holds a value of type `T` when $(I non-empty), and nothing when
+ * $(I empty). It is $(I empty) in the default initialized state.
+ *
+ * Construct with $(LREF none) for explicitly empty instances, $(LREF some)
+ * for non-empty instances and $(LREF option) to convert from nullable values.
+ *
+ * Option eliminates null dereferences in two ways. Primarily by encouraging
+ * range algorithms, which tend to have the desired behavior on empty ranges.
+ * For other cases, all access (dereference) to the underlying value is
+ * protected by explicit calls such as $(LREF _Option.getOrElse) and
+ * $(LREF _Option.assumeNonEmpty).
+ *
+ * For non-mutable types of `T`, Option can only be reassigned to $(LREF none).
+ * Params:
+ *    T = type of contained value
+ */
+struct Option(T)
+{
+    private T value = T.init; // For when T has @disable this();
+    private bool isDefined = false;
+
+    // Used by constructor functions
+    private this(T value, bool isDefined)
+    {
+        static if (isMutable!T)
+        {
+            import std.algorithm.mutation : move;
+            this.value = move(value);
+        }
+        else this.value = value;
+        this.isDefined = isDefined;
+    }
```

Comment:
Not sure this function is sound. If `isDefined` is false, there's no need to do anything.